### PR TITLE
Prelim/descriptions

### PR DESCRIPTION
**Pull request by andre-merzky**
This implements a fast description base class for the RCT stack which supports attribute based accessors and schema based type checking.  Ssee [here](https://github.com/radical-cybertools/radical.pilot/blob/prelim/descriptions/docs/architecture/dicts/README.md) for motivation and performance measurements.

@mturilli:  I asked @iparask for a review, but you may want to have a look, too, if you find the time....

### CODE REVIEW

File: src/radical/utils/description.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+
+    _verifier_keys = list(_verifiers.keys())
+
+    @classmethod
+    def _verify_kvt(cls, k, v, t):
+
+        if t is None              : return v
+        if t in cls._verifier_keys: return cls._verifiers[t](k, v, t)
+        if isinstance(t, dict)    : return cls._verify_dict(k, v, t)
+        if isinstance(t, list)    : return cls._verify_list(k, v, t)
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/utils/description.py]: I added this - but in general we should only add types as needed, as an additional `isinstance` incurs a (small) runtime penalty.  I should also mention that I find tuples confusing in Python:
```python
>>> l = []; print(type(l))
<class 'list'>
>>> l = [1]; print(type(l))
<class 'list'>
>>> l = [1, 2]; print(type(l))
<class 'list'>

>>> t = (); print(type(t))
<class 'tuple'>
>>> t = (1); print(type(t))
<class 'int'>
>>> t = (1, 2); print(type(t))
<class 'tuple'>
```

and so try avoid using them.  They don't have much benefit over lists (for us).

File: src/radical/utils/description.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        self[k] = v
+
+
+    # --------------------------------------------------------------------------
+    #
+    # don't list private class attributes (starting with `_`) as dict entries
+    #
+    def __iter__(self):
+        for k in dict.__iter__(self):
+            if str(k).startswith('_'):
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/utils/description.py]: This code is gone now - but thanks, you are right (char comparison is faster).

File: src/radical/utils/description.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+
+        from_dict: data to be used for initialization
+
+        NOTE: Keys containing an underscore are not exposed via the API.
+              Keys containing dots are split and interpreted as paths in te
+              configuration hierarchy.
+        '''
+
+        # TODO: `as_dict`, `keys` and `items` are reserved attributes.
+        #       Should this list be smaller?
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/utils/description.py]: There are now no more restrictions on usable keys - but those which collide with method names can only be used via the dict interface.  That is now documented in the docstring.

File: src/radical/utils/description.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+    def _verify_kvt(cls, k, v, t):
+
+        if t is None              : return v
+        if t in cls._verifier_keys: return cls._verifiers[t](k, v, t)
+        if isinstance(t, dict)    : return cls._verify_dict(k, v, t)
+        if isinstance(t, list)    : return cls._verify_list(k, v, t)
+        print()
+        print(t)
+        print(isinstance(t, list))
+        print(type(t))
```
[the remark this answers is not in the document]
Comment (andre-merzky) [reply src/radical/utils/description.py]: gone, thanks!

**Comment by andre-merzky** [comment]
@mturilli : thanks for the typo feedback! :-)

**Comment by andre-merzky** [comment]
Thanks for the feedback @mturilli !  Have another look at the [README](https://github.com/radical-cybertools/radical.pilot/blob/prelim/descriptions/docs/architecture/dicts/README.md), please - the RU implementation is now actually also the fastest for `deepcopy`.  Tests were added, too.

**Comment by andre-merzky** [comment]
Hi @mturilli : I think addressed all comments.  Wanna have another look?

File: src/radical/utils/description.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mturilli) [reply src/radical/utils/description.py]: Sounds good, thank you!

**Comment by mturilli** [comment]
README looks good!

**Pull request merged**